Ледянкин Павел

PavelLediankin opened the pull request.

PavelLediankin committed:
Перенесен проект TagsCloudVisualization, изменения к задаче

PavelLediankin committed:
Контейнер,

PavelLediankin committed:
Helper

(для внутреннего использования)

PavelLediankin committed:
Tests

PavelLediankin commented:
@HolyPrapor

HolyPrapor commented on TagCloudContainerTests/TTests.cs:
```diff
@@ -0,0 +1,18 @@
+        // И Layouter из контейнера, но тут не совсем понятно что тестировать
+        // (отличия от TagsCloudVisualization - почти отсутствуют),
+        // Может выделить класс - который дает список тэгов (название - размер), которые будут отправелны в TagsCloudVisualization?
+        // Кстати а нет смысла выделять проект- Парсер, чтобы контейнер занимался только своей задачей?
```
Давай на все вопросы вместе порассуждаем, можешь в тг написать чтобы быстрее обсудили)

1) Нужно ли нам столько проектов? В чём польза выделения чего-то в отдельный проект?
2) На какие части мы можем написать адекватные тесты?

PavelLediankin replied:
1. Думаю польза в том что каждый проект занимается одним своим делом (условно SRP), контейнер занимается контейнеризацией, парсер парсит, визуализатор рисует, Но если не надо выделять то я только за.

2.TagsCloudVisualization (из прошлой задачи с необходимыми изменениями и доп тестами)
Раскладка(из нового соответствие текст-размер, отсутствие повторений  т.д.) ну и визуализацию наверное не получится тестировать. 

Тестировать в контейнере -
Из Common не думаю что нужно, остается парсер(хотя опять же тесты тут будут довольно простыми и очевидными), и фильтр

HolyPrapor replied:
1. SRP это конечно хорошо, но это скорее применимо к классам. И нужен он в том числе для переиспользования кода, чтобы получилось более модульно.
С проектами точно так же: если что-то может понадобиться отдельно, вне другого кода - можно выделить в другой проект.
В твоём случае, я думаю, это немного чрезмерно. Другими словами: можешь выделить два проекта (тесты и не тесты), можно ещё отделить клиент (текстовый/gui) в отдельный проект, по желанию :)

2. Да, в целом всё так. Можем протестировать предобработку, подсчет статистики, фильтрацию, и окрашивание (про это подробнее в другом комментарии)

HolyPrapor commented on TagWriterHelper/Program.cs:
```diff
@@ -0,0 +1,21 @@
+    {
+        static void Main(string[] args)
+        {
+            //Helper - чтобы не заполнять файлы с тэгами вручную
```
Давай переназовём проект как "Local", подразумевая, что это какое-то локальное тестирование?
Можно ещё это перенести в Tests, тоже норм вариант думаю

Иметь проект с названием `Helper` как-то немного странно

PavelLediankin replied:
Вообще предполагалось, что я смогу записать тэги не только для тестов но и для основного проекта, но думаю в принципе можно перенести в тесты

HolyPrapor commented on TagWriterHelper/Tag.cs:
```diff
@@ -0,0 +1,22 @@
+
+namespace TagWriterHelper
+{
+    public class Tag
```
Этот класс, кажется, должен быть не в этом проекте, перенесем?

HolyPrapor commented on TagWriterHelper/Writer.cs:
```diff
@@ -0,0 +1,24 @@
+
+namespace TagWriterHelper
+{
+    public static class Writer
```
У тебя ведь уже есть класс `TagWriter`, чем этот от него отличается?

PavelLediankin replied:
Обертка над обычным writer-ом, которая получает список тэгов и записывает их правильно, 
обычный writer, может использоваться и для других целей, и не должен знать особенностей записи именно тэгов

HolyPrapor replied:
Тогда давай имя получше придумаем. `Writer` совсем не дает понимания, что за класс такой)

HolyPrapor commented on TagsCloudContainer/CloudVisualizator.cs:
```diff
@@ -0,0 +1,19 @@
+
+namespace TagsCloudContainer
+{
+    public class CloudVisualizator
```
Можешь рассказать, для чего нужен этот класс?

PavelLediankin replied:
Вообще странно, зачем я его выделил, похоже того же самого можно добиться просто выделив конструктор у визуализатора из другого проекта

HolyPrapor commented on TagsCloudContainer/Program.cs:
```diff
@@ -0,0 +1,91 @@
+                var wordsCount = parser.Parse(text);
+                var cloud = layouter.PlaceTagsInCloud(wordsCount, minSize, maxScale);
+                visualizator.Visualize(cloud);
+            }
```
Давай сконцентрируемся на этом методе, тут хорошо видно поток данных, молодец что явно выписал его)

Вопросы к расширяемости, которые у меня появились:
1) Что, если я захочу исключить какие-то конкретные слова? А если захочу исключить только существительные? А если захочу исключить слова оканчивающиеся на "й" в нормальной форме? Для этого есть точка расширения?
2) Кажется, нет точки для предобработки тегов: приведение в нижний регистр, в нормальную форму, ...
3) Предположим, мы умеем определять часть речи. Как мы можем реализовать фичу "хочу окрашивать существительные в синий, а прилагательные в жёлтый"?

PavelLediankin replied:
Выделить один  класс для предобработки и исключения (фильтрация) или два?
 Для исключений напрямую записать что исключать или использовать селектор?

3. С тем что у меня есть это будет неудобно добавить, можно тэгу добавить enum часть речи и по нему задавать цвет тэга. Но тогда вопрос есть ли смысл оставлять текущую установку цвета по номеру цвета в переданном списке ?

HolyPrapor replied:
А сам как думаешь, лучше один класс или два?

3. Будет не очень, в том плане, что не всегда у нас в предобработке будет "анализатор части речи", и будет странно, если мы им не пользуемся, а в настройки должны передавать что-то на эту тему.
Как насчёт такого варианта: после предобработок и подсчёта статистики, перед непосредственно визуализацией, добавить новый класс "TagPainter", который бы мог принимать обычные теги, а возвращать "окрашенные"?
Тогда визуализатор просто брал их, и рисовал согласно их цвету. Что думаешь?

PavelLediankin replied:
Думаю лучше два.
3. Да, думаю это не сложно будет сделать

HolyPrapor replied:
А почему лучше два?

Опиши интерфейсы с одним и с двумя классами, и объясни какой выбираешь и почему

HolyPrapor commented on TagsCloudContainer/WordsCountParser.cs:
```diff
@@ -0,0 +1,25 @@
+            return wordsCount;
+        }
+
+        private IEnumerable<string> GetWordsWithoutUseless(IEnumerable<string> words)
```
Кажется, в этом классе смешалась фильтрация и подсчет количества тегов.

Подумай как это можно исправить с учётом моего предыдущего комментария)

HolyPrapor commented on di.sln:
```diff
@@ -1,6 +1,17 @@
+# Visual Studio Version 16
+VisualStudioVersion = 16.0.31112.23
+MinimumVisualStudioVersion = 10.0.40219.1
+Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "FractalPainter", "FractalPainter\FractalPainter.csproj", "{4D70883B-6F8B-4166-802F-8EDC9BE93199}"
```
Выпилим все лишние проекты из солюшена?

PavelLediankin committed:
убрал CloudVisualizator

PavelLediankin committed:
перенес TagsWriterHelper в тесты

PavelLediankin committed:
перенес класс простого тэга в проект контейнер

PavelLediankin committed:
Разбиты фильтрация, нормализация и парсинг

PavelLediankin committed:
Tag painter

PavelLediankin committed:
Исправлены небольшие ошибки

HolyPrapor commented on TagsCloudContainer/Program.cs:
```diff
@@ -0,0 +1,100 @@
+                var text = reader.Read(textFilename);
+                var tags = parser.Parse(text);
+                // Подумаю еще как это можно сделать лучше >
+                tags = new Normalizator().Process(new TagsFilter().Process(tags)).ToList();
```
Вижу, что в итоге решил оставить только `IPreprocessor`.

По поводу этой заметки: может просто выделить класс `WordPreprocessor`, который бы на вход принимал список препроцессоров, и просто его применял?
Типа такой контейнер для препроцессоров

PavelLediankin replied:
Да, как раз что-то подобное я и собирался сделать

HolyPrapor commented on TagsCloudContainer/Program.cs:
```diff
@@ -0,0 +1,100 @@
+                var tags = parser.Parse(text);
+                // Подумаю еще как это можно сделать лучше >
+                tags = new Normalizator().Process(new TagsFilter().Process(tags)).ToList();
+                var paintedTags = painter.Paint(tags);
```
Лично меня немного смущает, что теги сначала окрашиваются, и только потом расставляются.

В таком виде не получится сделать какой-нибудь красивый градиент цвета тега в зависимости от позиции относительно центра облака. Может поменяем порядок? 

Итоговый порядок я вижу как-то так:
1. Чтение
2. Парсинг в теги
3. Препроцессинг
4. Подсчёт статистики
5. Расположение тегов в прямоугольники (раскладка)
6. Окрашивание
7. Отрисовка

Что думаешь?

HolyPrapor replied:
У тебя тогда ещё уйдет протечка абстракции: в `Factory` у `ITag` проник параметр `Palette`, что кажется не очень правильным

PavelLediankin replied:
Ну не знаю, для градиента это и впрямь будет удобнее, но если нужно  привязать  цвета к частям речи, то не так удобно. Но думаю в принципе это возможно.
Кстати, а в чем протечка ? Разве лучше передавать отдельно цвета?

HolyPrapor replied:
> Но если нужно привязать цвета к частям речи, то не так удобно.

А в чём неудобство заключается? На этапе окраски у тебя есть слово, можно его прямо в этот момент и проанализировать на часть речи.

Вся эта мета-информация может вычисляться на месте, не стоит пытаться ускорить код (например, выполняя проверку части речи лишь один раз, а затем таскать эту инфу за собой) в угоду читаемости и простоты. 

Тем более, что операция определения части речи не такая уж и дорогая, а наше приложение не сильно требовательное к производительности :)

> Кстати, а в чем протечка ? Разве лучше передавать отдельно цвета?

Я не про цвета. 
Протечка в том, что тебе при желании добавить что-то в конструктор тега придется постоянно добавлять это в `Factory` у `ITag`, и таскать эти параметры за собой. Мне кажется, что это неудобно(
Тем более, когда есть вариант этого не делать.
Сначала у тебя палитра проникла, потом часть речи, потом ещё какая-нибудь информация.

HolyPrapor commented on TagsCloudContainer/Preprocessors/TagsFilter.cs:
```diff
@@ -0,0 +1,18 @@
+
+namespace TagsCloudContainer
+{
+    public class TagsFilter : IPreprocessor
```
Подберешь потом название получше, когда фильтрация поумнее станет?

PavelLediankin replied:
Теперь лучше?

HolyPrapor replied:
Ага, так хорошо видна суть :)

HolyPrapor commented on TagsCloudContainer/Preprocessors/Normalizator.cs:
```diff
@@ -0,0 +1,20 @@
+
+namespace TagsCloudContainer
+{
+    public class Normalizator : IPreprocessor
```
Нормализация может много шагов в себя включать, не хочется их все сюда тащить, так что можно просто назвать `ToLower` или `ToLowerPreprocessor`

HolyPrapor commented on TagsCloudContainer/Painting/TagPainter.cs:
```diff
@@ -0,0 +1,29 @@
+            this.maker = maker;
+        }
+
+        public List<SimpleTag> Paint(List<SimpleTag> tags)
```
Может быть стоит здесь разделить абстракции?
Принимать `SimpleTag`, а возвращать `PaintedTag`?

PavelLediankin replied:
Если оставлять порядок покраски, то бессмысленно, к той же информации добавляется только информация о  цвете, Если красить после раскладки то PaintedTag вообще не нужен

HolyPrapor commented on TagsCloudContainer/Painting/TagPainter.cs:
```diff
@@ -0,0 +1,29 @@
+
+        public List<SimpleTag> Paint(List<SimpleTag> tags)
+        {
+            var enumerator = maker.GetPalettes().GetEnumerator();
```
Возможно, тебе помог бы метод `.Zip`

HolyPrapor commented on TagsCloudContainer/Layouters/TagCircularLayouter.cs:
```diff
@@ -0,0 +1,28 @@
+            foreach (var simpleTag in tags)
+            {
+                var size = GetSize(simpleTag.Word, minHeight, maxScale, simpleTag.Count);
+                var tag = factory(simpleTag.Word, simpleTag.Palette);
```
Вот не очень прикольно, что сюда какая-то палитра проникла :(

Писал об этом в комментарии выше

HolyPrapor commented on TagsCloudContainer/Common/SimpleTag.cs:
```diff
@@ -0,0 +1,24 @@
+    {
+        public readonly string Word;
+        public readonly int Count;
+        public Palette  Palette { get; set; }
```
```suggestion
        public Palette Palette { get; set; }
```
Не забывай форматировать код перед отправкой)

PavelLediankin committed:
Zip для двойного перечисления

PavelLediankin committed:
refactoring

PavelLediankin committed:
Preprocessors update

PavelLediankin committed:
Отрисовка после раскладки

+ добавлена палитра по умолчанию, на случай если забудут о TagPainter

PavelLediankin committed:
Тесты

+ небольшие изменения исправляющие мелкие ошибки, упрощение тестирования

PavelLediankin committed:
Предобработка слов

дать возможность влиять на список скучных слов, которые не попадут в облако. (Список скучных слов или селектор)

PavelLediankin committed:
Алгоритм

реализовать несколько алгоритмов формирования облака тегов. (крупные тэги собраны в центре или по краям)

PavelLediankin committed:
Добавлен атрибут для препроцессоров

Позволяет добавить(или убрать процессоры в настройках)

PavelLediankin committed:
Меню

PavelLediankin committed:
Ui

+ изменения в процессе создания и отладки Ui

PavelLediankin committed:
убрал лишние юзинги

PavelLediankin committed:
Меню и переименование ResolveContainer

PavelLediankin committed:
Рефактор, исправление ошибок, убрал атрибуты

PavelLediankin committed:
Добавлен ConsoleUiAction, немного упрощено создание второго клиента
